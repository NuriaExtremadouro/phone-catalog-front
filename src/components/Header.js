import React from 'react';
import { Grid } from 'semantic-ui-react';
import { Link } from "react-router-dom";

/**
 * Top menu that is always placed at the top of the screen.
 */
export default () => (
    <Grid columns='equal' verticalAlign="middle" className="top-header">
        <Grid.Column textAlign="center">
            <Link to='/'>
            Phone Catalog
            <div className="underline"></div>
            </Link>
        </Grid.Column>
    </Grid>
)
