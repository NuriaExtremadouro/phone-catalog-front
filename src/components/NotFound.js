import React from 'react';
import { Grid, Header } from 'semantic-ui-react';
import { Link } from "react-router-dom";

/**
 * Page to show in case the user attepts to navigate to a non-existent page.
 */
export default () => (
    <div className="phone-list-container">
        <Grid columns="equal" textAlign="center">
            <Grid.Column>
                <Header as="h1">Page not found</Header>
                <Link to="/">Go back</Link>
            </Grid.Column>
        </Grid>
    </div>
)
