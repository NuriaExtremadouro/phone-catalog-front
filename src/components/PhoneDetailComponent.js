import React, { Component } from 'react';
import { Grid, Dimmer, Loader, Image, Segment, Divider, Header } from 'semantic-ui-react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router';
import { getPhone } from '../actions/phoneActions';

/**
 * Component that shows the details of a specific phone.
 */
export class PhoneDetailComponent extends Component {
    /**
     * Retrieve the phone data when the component loads.
     */
    componentDidMount() {
        this.props.getPhone(parseInt(this.props.match.params.id));
    }

    /**
     * Renders the view.
     */
    render() {
        // If there was a problem when retrieving the phone, it redirects to the NotFound page.
        if (!this.props.phone) {
            return <Redirect to='/notfound' />
        }

        return (
            <div className="phone-list-container">
                {this.props.isFetching ? 
                    <Dimmer active><Loader>Loading</Loader></Dimmer> :
                    <Grid>
                        <Segment className="details-data">
                            <Grid columns="equal">
                                <Grid.Column mobile={16} computer={8}>
                                    <Image className="details-image" src={this.props.phone.image}/>
                                </Grid.Column>
                                <Grid.Column mobile={16} computer={8}>
                                    <Header as='h1' dividing>{this.props.phone.name}</Header>
                                    <Grid columns="equal">
                                        <Grid.Column mobile={16} computer={8}>
                                            <label>Screen: </label>
                                            <p>{this.props.phone.screen}</p>
                                            <label>RAM: </label>
                                            <p>{this.props.phone.ram}</p>
                                            <label>Internal Memory: </label>
                                            <p>{this.props.phone.internalMemory}</p>
                                            <label>Camera: </label>
                                            <p>{this.props.phone.camera}</p>
                                        </Grid.Column>
                                        <Grid.Column mobile={16} computer={8}>
                                            <label>Micro SD: </label>
                                            <p>{this.props.phone.microSD ? "Yes" : "No"}</p>
                                            <label>Battery: </label>
                                            <p>{this.props.phone.battery}</p>
                                            <label>Color: </label>
                                            <p>{this.props.phone.color}</p>
                                            <label>Price: </label>
                                            <p>{this.props.phone.price} â‚¬</p>
                                        </Grid.Column>
                                    </Grid>
                                    <Divider horizontal>Description</Divider>
                                    <p>{this.props.phone.description}</p>
                                </Grid.Column>
                            </Grid>
                        </Segment>
                    </Grid>
                }
            </div>
        );
    }
}

PhoneDetailComponent.propTypes = {
    getPhone: PropTypes.func.isRequired,
    phone: PropTypes.object.isRequired,
    isFetching: PropTypes.bool.isRequired
}

const mapStateToProps = state => ({
  phone: state.data.phone,
  isFetching: state.data.isFetching
});

export default connect(mapStateToProps, { getPhone })(PhoneDetailComponent);