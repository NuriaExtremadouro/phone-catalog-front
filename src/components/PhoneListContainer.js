import React, { Component } from 'react';
import { Dimmer, Grid, Icon, Image, Loader, Header, Segment } from 'semantic-ui-react';
import { Link } from "react-router-dom";
import { connect } from 'react-redux';
import { getPhones } from '../actions/phoneActions';
import { PropTypes } from 'prop-types';

/**
 * Component that shows the list of phones.
 */
export class PhoneListContainer extends Component {
  /**
   * Retrieves the phone list when the component loads.
   */
  componentDidMount() {
    this.props.getPhones();
  }

  /**
   * Renders the view.
   */
  render() {
    return (
      <div className="phone-list-container">
        {this.props.isFetching ?
          <Dimmer active><Loader>Loading</Loader></Dimmer> :
          <Grid>
            {this.props.phones.map((phone) =>
              <Grid.Column key={phone.id} widescreen={4} largeScreen={5} mobile={16}>
                <Link to={`/detail/${phone.id}`}>
                  <Segment className="phone-card">
                    <Image className="card-image" src={phone.image} />
                    <div className="card-text">
                      <Header as='h3'>{phone.name}</Header>
                      <p>{phone.price} â‚¬</p>
                      <Header as='h4' className="card-details">View details<Icon name="angle right"></Icon></Header>
                    </div>
                  </Segment>
                </Link>
              </Grid.Column>
            )}
          </Grid>
        }
      </div>
    );
  }
}

PhoneListContainer.propTypes = {
  getPhones: PropTypes.func.isRequired,
  phones: PropTypes.array.isRequired,
  isFetching: PropTypes.bool.isRequired
}

const mapStateToProps = state => ({
  phones: state.data.phones,
  isFetching: state.data.isFetching
});

export default connect(mapStateToProps, { getPhones })(PhoneListContainer)