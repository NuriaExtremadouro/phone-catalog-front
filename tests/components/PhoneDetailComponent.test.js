import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PhoneDetailComponent } from '../../src/components/PhoneDetailComponent';
import { Image, Header } from 'semantic-ui-react';

describe('PhoneDetailComponent', () => {
    // Wrapper that contains an instance of the component that is being tested.
    let wrapper;
    // Mock functions and data to simulate the store.
    const mockGetPhone = jest.fn(Number);
    const mockPhone = {
        "id": 1,
        "name": "Xiaomi Redmi 5 Plus",
        "description": "No hay año en que Xiaomi domine con contundencia la lista de mejores teléfonos de menos de 200 euros si lo que buscamos es lo máximo por nuestro dinero. El Xiaomi Redmi 5 Plus es una garantía de éxito que nos da potencia, batería y mucha memoria interna contando además con una gran pantalla.",
        "screen": "6'",
        "ram": "4 GB",
        "internalMemory": "64 GB",
        "camera": "12 MP",
        "microSD": true,
        "battery": "4000 mAh",
        "color": "blue",
        "price": 181,
        "rating": 5,
        "featured": true,
        "image": "https://cdn2.gsmarena.com/vv/pics/xiaomi/xiaomi-redmi-5-plus-1.jpg"
    };
    
    beforeEach(() => {
        wrapper = shallow(<PhoneDetailComponent getPhone={mockGetPhone} phone={mockPhone} isFetching={false} match={{params: {id: 1}, isExact: true, path: "", url: ""}}/>)
    });

    it('renders the component', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('gets the phone data on componentDidMount', () => {
        expect(mockGetPhone.mock.calls.length).toBe(1);
    });

    it('shows the data of the phone', () => {
        expect(wrapper.find(Image).props().src).toEqual(mockPhone.image);
        expect(wrapper.find(Header).props().children).toEqual(mockPhone.name);
        expect(wrapper.find('p').at(0).props().children).toEqual(mockPhone.screen);
        expect(wrapper.find('p').at(1).props().children).toEqual(mockPhone.ram);
        expect(wrapper.find('p').at(2).props().children).toEqual(mockPhone.internalMemory);
        expect(wrapper.find('p').at(3).props().children).toEqual(mockPhone.camera);
        expect(wrapper.find('p').at(4).props().children).toEqual(mockPhone.microSD ? "Yes" : "No");
        expect(wrapper.find('p').at(5).props().children).toEqual(mockPhone.battery);
        expect(wrapper.find('p').at(6).props().children).toEqual(mockPhone.color);
        expect(wrapper.find('p').at(7).props().children[0]).toEqual(mockPhone.price);
        expect(wrapper.find('p').at(8).props().children).toEqual(mockPhone.description);
    });
})