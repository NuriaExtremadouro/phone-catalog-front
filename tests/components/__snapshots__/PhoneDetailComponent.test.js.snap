// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneListContainer renders the component 1`] = `
<div
  className="phone-list-container"
>
  <Grid>
    <Segment
      className="details-data"
    >
      <Grid
        columns="equal"
      >
        <GridColumn
          computer={8}
          mobile={16}
        >
          <Image
            as="img"
            className="details-image"
            src="https://cdn2.gsmarena.com/vv/pics/xiaomi/xiaomi-redmi-5-plus-1.jpg"
            ui={true}
          />
        </GridColumn>
        <GridColumn
          computer={8}
          mobile={16}
        >
          <Header
            as="h1"
            dividing={true}
          >
            Xiaomi Redmi 5 Plus
          </Header>
          <Grid
            columns="equal"
          >
            <GridColumn
              computer={8}
              mobile={16}
            >
              <label>
                Screen: 
              </label>
              <p>
                6'
              </p>
              <label>
                RAM: 
              </label>
              <p>
                4 GB
              </p>
              <label>
                Internal Memory: 
              </label>
              <p>
                64 GB
              </p>
              <label>
                Camera: 
              </label>
              <p>
                12 MP
              </p>
            </GridColumn>
            <GridColumn
              computer={8}
              mobile={16}
            >
              <label>
                Micro SD: 
              </label>
              <p>
                Yes
              </p>
              <label>
                Battery: 
              </label>
              <p>
                4000 mAh
              </p>
              <label>
                Color: 
              </label>
              <p>
                blue
              </p>
              <label>
                Price: 
              </label>
              <p>
                181
                 €
              </p>
            </GridColumn>
          </Grid>
          <Divider
            horizontal={true}
          >
            Description
          </Divider>
          <p>
            No hay año en que Xiaomi domine con contundencia la lista de mejores teléfonos de menos de 200 euros si lo que buscamos es lo máximo por nuestro dinero. El Xiaomi Redmi 5 Plus es una garantía de éxito que nos da potencia, batería y mucha memoria interna contando además con una gran pantalla.
          </p>
        </GridColumn>
      </Grid>
    </Segment>
  </Grid>
</div>
`;
